detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GCPMetadataConfig.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2021 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Retrieves Google Cloud project-id and a limited set of instance attributes from Metadata server.
+ *
+ * @see <a href="https://cloud.google.com/compute/docs/storing-retrieving-metadata">
+ *     https://cloud.google.com/compute/docs/storing-retrieving-metadata</a>
+ */
+final class GCPMetadataConfig {
+  private static final String DEFAULT_URL = "http://metadata.google.internal/computeMetadata/v1/";
+  public static final GCPMetadataConfig DEFAULT_INSTANCE = new GCPMetadataConfig(DEFAULT_URL);
+
+  private final String url;
+
+  // For testing only
+  public GCPMetadataConfig(String url) {
+    this.url = url;
+  }
+
+  boolean isRunningOnGcp() {
+    return !getProjectId().isEmpty();
+  }
+
+  String getProjectId() {
+    return getAttribute("project/project-id");
+  }
+
+  String getZone() {
+    String zone = getAttribute("instance/zone");
+    if (zone.contains("/")) {
+      return zone.substring(zone.lastIndexOf('/') + 1);
+    }
+    return zone;
+  }
+
+  String getMachineType() {
+    String machineType = getAttribute("instance/machine-type");
+    if (machineType.contains("/")) {
+      return machineType.substring(machineType.lastIndexOf('/') + 1);
+    }
+    return machineType;
+  }
+
+  String getInstanceId() {
+    return getAttribute("instance/id");
+  }
+
+  String getClusterName() {
+    return getAttribute("instance/attributes/cluster-name");
+  }
+
+  String getInstanceHostName() {
+    return getAttribute("instance/hostname");
+  }
+
+  String getInstanceName() {
+    return getAttribute("instance/name");
+  }
+
+  String getAttribute(String attributeName) {
+    try {
+      URL url = new URL(this.url + attributeName);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestProperty("Metadata-Flavor", "Google");
+      InputStream input = connection.getInputStream();
+      if (connection.getResponseCode() == 200) {
```

Comment:
FYI - This doesn't actually mean you got a metadataserver response, some kind of metadata response validation is needed.  E.g. many places DNS squat and hijack non-existent DNS lookups with ad-server homepages.  We have an open bug about this and Verizon FIOS users.

Can you add a TODO to validate that the response actually came from GCP metadata server?

Reply:
I'll work out the details on the metadata header with the team and describe the use case.  Ideally we can set up an integration test that forces them to exist.  I sent you the internal bug to track details here.

I'm ok with any kind of checking, but the suggestion now is to use the metadata tag.   We'll shore this up, but we'll also need to do a similar fix to other SDKs.    For now, let's use the metadata http header (because it's very easy to check) and offline (and outside the CL) we'll work on other validation mechanisms.